File: apps/visualizer/backend/api/schemas/identity.py
"""Identity API response models — best photos, mirror signature, post-next suggestions."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict

from api.schemas.catalog import IdentityPerPerspectiveScore

ImageTypeLiteral = Literal['catalog', 'instagram']


class IdentityBestPhotoItem(BaseModel):
    model_config = ConfigDict(extra='forbid')

    image_key: str
    image_type: ImageTypeLiteral | None = None
    peak_percentile: float
    perspectives_covered: int
    eligible: bool | None = None
    per_perspective: list[IdentityPerPerspectiveScore]
    filename: str
    date_taken: str
    rating: int
    instagram_posted: bool
    stack_id: int | None = None
    stack_member_count: int | None = None
    is_stack_representative: bool | None = None


class IdentityBestPhotosMeta(BaseModel):
    model_config = ConfigDict(extra='forbid')

    active_perspectives: list[str] | None = None
    weighting: str | None = None
    ranking_key: str | None = None
    min_perspectives_used: int | None = None
    coverage_rule: str | None = None
    total_catalog_images: int | None = None
    eligible_count: int | None = None
    scored_any_count: int | None = None
    coverage_note: str | None = None


class IdentityBestPhotosResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    items: list[IdentityBestPhotoItem]
    total: int
    meta: IdentityBestPhotosMeta


class IdentityBestPhotosQuery(BaseModel):
    limit: int | None = None
    offset: int | None = None
    min_perspectives: int | None = None
    sort_by_date: str | None = None
    posted: str | None = None


class MirrorDescriptor(BaseModel):
    model_config = ConfigDict(extra='forbid')

    token: str
    log_odds: float
    count: int


class MirrorExemplarPerPerspective(BaseModel):
    model_config = ConfigDict(extra='forbid')

    perspective_slug: str
    display_name: str
    score: int
    percentile: float


class MirrorExemplar(BaseModel):
    model_config = ConfigDict(extra='forbid')

    image_key: str
    filename: str
    date_taken: str
    rating: int
    instagram_posted: bool
    score: int
    percentile: float
    purity: float
    rationale_preview: str
    per_perspective: list[MirrorExemplarPerPerspective]
    stack_id: int | None = None
    stack_size: int | None = None


class MirrorTechniqueSection(BaseModel):
    model_config = ConfigDict(extra='forbid')

    perspective_slug: str
    display_name: str
    strength_label: str
    leading_not_distinctive: bool
    crowned: bool
    win_rate: float
    chance_rate: float
    z_score: float
    votes: int
    photos_on: int
    coverage: float
    low_coverage: bool
    descriptors: list[MirrorDescriptor]
    exemplars: list[MirrorExemplar]
    exemplar_total: int


class MirrorOtherLens(BaseModel):
    model_config = ConfigDict(extra='forbid')

    perspective_slug: str
    display_name: str
    strength_label: str
    win_rate: float
    chance_rate: float
    z_score: float
    coverage: float
    low_coverage: bool
    votes: int
    photos_on: int
    exemplar_total: int


class MirrorMeta(BaseModel):
    model_config = ConfigDict(extra='forbid')

    active_perspectives: list[str] | None = None
    total_catalog_images: int | None = None
    voting_rule: str | None = None
    crowning_rule: str | None = None
    low_coverage_threshold: float | None = None
    exemplar_initial_limit: int | None = None
    exemplar_page_size: int | None = None
    descriptor_min_count: int | None = None
    scores_are_advisory: str | None = None
    fallback_active: bool | None = None


class MirrorResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    population: int
    sections: list[MirrorTechniqueSection]
    other_lenses: list[MirrorOtherLens]
    meta: MirrorMeta


class MirrorLensExemplarsResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    items: list[MirrorExemplar]
    total: int


class MirrorLensExemplarsQuery(BaseModel):
    limit: int | None = None
    offset: int | None = None


class PostNextCandidate(BaseModel):
    model_config = ConfigDict(extra='forbid')

    image_key: str
    image_type: ImageTypeLiteral | None = None
    filename: str
    date_taken: str
    rating: int
    peak_percentile: float
    peak_perspective_slug: str
    peak_perspective_display_name: str
    is_signature: bool
    perspectives_covered: int
    per_perspective: list[IdentityPerPerspectiveScore]
    reasons: list[str]
    reason_codes: list[str]


class PostNextSuggestionsMeta(BaseModel):
    model_config = ConfigDict(extra='forbid')

    weighting: str | None = None
    ranking_key: str | None = None
    min_perspectives_used: int | None = None
    coverage_rule: str | None = None
    high_score_rule: str | None = None


class PostNextSuggestionsResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    candidates: list[PostNextCandidate]
    total: int
    meta: PostNextSuggestionsMeta
    empty_state: str | None = None


class PostNextSuggestionsQuery(BaseModel):
    limit: int | None = None
    offset: int | None = None
    sort_by_date: str | None = None

File: apps/visualizer/backend-ts/src/api/schemas/identity.ts
/**
 * Identity API response models. Port of `api/schemas/identity.py`.
 *
 * One deliberate difference from the Python models, and it is a bug fix rather than
 * a port: `IdentityBestPhotoItem` gains `ranking_percentile`,
 * `corroboration_revoked` and `corroboration_revoked_by`.
 *
 * The corroboration veto (#292, commit f8b6662) started returning those three
 * fields from `rank_best_photos`, but the pydantic model sets `extra='forbid'` and
 * was never updated — so spectree's response validation rejects the payload and
 * `GET /api/identity/best-photos` answers **500** on the real catalog. Verified
 * against the running Flask app. Reproducing that would mean porting a broken page,
 * so the fields are declared here and the endpoint works. Nothing in the frontend
 * references them yet, so the addition is purely additive.
 */
import { z } from '@hono/zod-openapi';
import { IdentityPerPerspectiveScore } from './catalog.js';

export const IdentityBestPhotoItem = z
  .object({
    image_key: z.string(),
    image_type: z.literal('catalog').nullish(),
    peak_percentile: z.number(),
    ranking_percentile: z.number(),
    corroboration_revoked: z.boolean(),
    corroboration_revoked_by: z.string(),
    perspectives_covered: z.int(),
    eligible: z.boolean().nullish(),
    per_perspective: z.array(IdentityPerPerspectiveScore),
    filename: z.string(),
    date_taken: z.string(),
    rating: z.int(),
    instagram_posted: z.boolean(),
    stack_id: z.int().nullish(),
    stack_member_count: z.int().nullish(),
    is_stack_representative: z.boolean().nullish(),
  })
  .strict()
  .openapi('IdentityBestPhotoItem');

export const IdentityBestPhotosMeta = z
  .object({
    active_perspectives: z.array(z.string()).nullish(),
    weighting: z.string().nullish(),
    ranking_key: z.string().nullish(),
    corroboration_rule: z.string().nullish(),
    min_perspectives_used: z.int().nullish(),
    coverage_rule: z.string().nullish(),
    total_catalog_images: z.int().nullish(),
    eligible_count: z.int().nullish(),
    scored_any_count: z.int().nullish(),
    coverage_note: z.string().nullish(),
  })
  .strict()
  .openapi('IdentityBestPhotosMeta');

export const IdentityBestPhotosResponse = z
  .object({
    items: z.array(IdentityBestPhotoItem),
    total: z.int(),
    meta: IdentityBestPhotosMeta,
  })
  .strict()
  .openapi('IdentityBestPhotosResponse');

export const MirrorDescriptor = z
  .object({ token: z.string(), log_odds: z.number(), count: z.int() })
  .strict()
  .openapi('MirrorDescriptor');

export const MirrorExemplarPerPerspective = z
  .object({
    perspective_slug: z.string(),
    display_name: z.string(),
    score: z.int(),
    percentile: z.number(),
  })
  .strict()
  .openapi('MirrorExemplarPerPerspective');

export const MirrorExemplar = z
  .object({
    image_key: z.string(),
    filename: z.string(),
    date_taken: z.string(),
    rating: z.int(),
    instagram_posted: z.boolean(),
    score: z.int(),
    percentile: z.number(),
    purity: z.number(),
    rationale_preview: z.string(),
    per_perspective: z.array(MirrorExemplarPerPerspective),
    stack_id: z.int().nullish(),
    stack_size: z.int().nullish(),
  })
  .strict()
  .openapi('MirrorExemplar');

export const MirrorTechniqueSection = z
  .object({
    perspective_slug: z.string(),
    display_name: z.string(),
    strength_label: z.string(),
    leading_not_distinctive: z.boolean(),
    crowned: z.boolean(),
    win_rate: z.number(),
    chance_rate: z.number(),
    z_score: z.number(),
    votes: z.int(),
    photos_on: z.int(),
    coverage: z.number(),
    low_coverage: z.boolean(),
    descriptors: z.array(MirrorDescriptor),
    exemplars: z.array(MirrorExemplar),
    exemplar_total: z.int(),
  })
  .strict()
  .openapi('MirrorTechniqueSection');

export const MirrorOtherLens = z
  .object({
    perspective_slug: z.string(),
    display_name: z.string(),
    strength_label: z.string(),
    win_rate: z.number(),
    chance_rate: z.number(),
    z_score: z.number(),
    coverage: z.number(),
    low_coverage: z.boolean(),
    votes: z.int(),
    photos_on: z.int(),
    exemplar_total: z.int(),
  })
  .strict()
  .openapi('MirrorOtherLens');

export const MirrorMeta = z
  .object({
    active_perspectives: z.array(z.string()).nullish(),
    total_catalog_images: z.int().nullish(),
    voting_rule: z.string().nullish(),
    crowning_rule: z.string().nullish(),
    low_coverage_threshold: z.number().nullish(),
    exemplar_initial_limit: z.int().nullish(),
    exemplar_page_size: z.int().nullish(),
    descriptor_min_count: z.int().nullish(),
    scores_are_advisory: z.string().nullish(),
    fallback_active: z.boolean().nullish(),
  })
  .strict()
  .openapi('MirrorMeta');

export const MirrorResponse = z
  .object({
    population: z.int(),
    sections: z.array(MirrorTechniqueSection),
    other_lenses: z.array(MirrorOtherLens),
    meta: MirrorMeta,
  })
  .strict()
  .openapi('MirrorResponse');

export const MirrorLensExemplarsResponse = z
  .object({ items: z.array(MirrorExemplar), total: z.int() })
  .strict()
  .openapi('MirrorLensExemplarsResponse');

export const PostNextCandidate = z
  .object({
    image_key: z.string(),
    image_type: z.literal('catalog').nullish(),
    filename: z.string(),
    date_taken: z.string(),
    rating: z.int(),
    peak_percentile: z.number(),
    peak_perspective_slug: z.string(),
    peak_perspective_display_name: z.string(),
    is_signature: z.boolean(),
    perspectives_covered: z.int(),
    per_perspective: z.array(IdentityPerPerspectiveScore),
    reasons: z.array(z.string()),
    reason_codes: z.array(z.string()),
  })
  .strict()
  .openapi('PostNextCandidate');

export const PostNextSuggestionsMeta = z
  .object({
    weighting: z.string().nullish(),
    ranking_key: z.string().nullish(),
    corroboration_rule: z.string().nullish(),
    min_perspectives_used: z.int().nullish(),
    coverage_rule: z.string().nullish(),
    high_score_rule: z.string().nullish(),
  })
  .strict()
  .openapi('PostNextSuggestionsMeta');

export const PostNextSuggestionsResponse = z
  .object({
    candidates: z.array(PostNextCandidate),
    total: z.int(),
    meta: PostNextSuggestionsMeta,
    empty_state: z.string().nullish(),
  })
  .strict()
  .openapi('PostNextSuggestionsResponse');
